src/sage/matroids/circuits_matroid.pyx

```diff
@@ -0,0 +1,676 @@
+            False
+        """
+        if memo is None:
+            memo = {}
+        from copy import deepcopy
+        # Since matroids are immutable, N cannot reference itself in correct code, so no need to worry about the recursion.
+        N = CircuitsMatroid(groundset=deepcopy(self._groundset, memo), circuits=deepcopy(frozenset(self._C), memo))
+        N.rename(deepcopy(self.get_custom_name(), memo))
+        return N
+
+    def __reduce__(self):
+        """
+        Save the matroid for later reloading.
+
+        OUTPUT:
+
+        A tuple ``(unpickle, (version, data))``, where ``unpickle`` is the
+        name of a function that, when called with ``(version, data)``,
+        produces a matroid isomorphic to ``self``. ``version`` is an integer
+        (currently 0) and ``data`` is a tuple ``(E, C, name)`` where ``E`` is
+        the groundset, ``C`` is the list of circuits, and ``name`` is a custom
+        name.
+
+        EXAMPLES::
+
+            sage: M = matroids.Theta(5)
+            sage: M == loads(dumps(M))  # indirect doctest
+            True
+            sage: M.reset_name()
+            sage: loads(dumps(M))
+            Matroid of rank 5 on 10 elements with 45 circuits
+        """
+        import sage.matroids.unpickling
+        data = (self._groundset, frozenset(self._C), self.get_custom_name())
+        version = 0
+        return sage.matroids.unpickling.unpickle_circuits_matroid, (version, data)
+
+    # enumeration
+
+    cpdef bases(self) noexcept:
+        r"""
+        Return the bases of the matroid.
+
+        OUTPUT: a SetSystem
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+            sage: len(M.bases())
+            6
+        """
+        cdef SetSystem B
+        B = SetSystem(list(self.groundset()))
+        from itertools import combinations
+        for S in combinations(self._groundset, self._matroid_rank):
+            flag = True
+            for C in self.nonspanning_circuits():
+                if C <= set(S):
+                    flag = False
+                    break
+            if flag:
+                B.append(S)
+        return B
+
+    def bases_iterator(self):
+        r"""
+        Return an iterator over the bases of the matroid.
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+            sage: it = M.bases_iterator()
+            sage: it.__next__()
+            frozenset({0, 1})
+        """
+        from itertools import combinations
+        for B in combinations(self._groundset, self._matroid_rank):
+            flag = True
+            for C in self.nonspanning_circuits():
+                if C <= set(B):
+                    flag = False
+                    break
+            if flag:
+                yield frozenset(B)
```

Comment:
```suggestion
        cdef SetSystem NSC = self.nonspanning_circuits()
        for B in combinations(self._groundset, self._matroid_rank):
            B = frozenset(B)
            if not any(C <= B for C in NSC)
                yield B
```
For normal `def` functions you can use generators IIRC. Anyways, it is best to not recreate things many times over.